cypress/support/PersistentLogs.js

```diff
@@ -16,15 +21,98 @@ console.log = function (message) {
   logEntries += message + '\r\n'
 }
 
-// This will be called by the test list manager to register each test case
-// in the order the tests will be run in.
-Cypress.PersistentLogs = {}
-Cypress.PersistentLogs.RegisterTestCase = (testGroupName, testName) => {
-  testCaseRegistry.push(`${testGroupName}/${testName}`)
-}
+// Write ONLY to the true console.log and NOT our persistent log.
+// Uncomment to use this for debugging purposes.
+// function consoleLogOnly() { originalConsolLog.apply(console, arguments) }
+
+// -----------------------------------------------------------------------------
+// Functions that add additional log entries related to the stage of the test 
+// suite and to persist the captured logs at the appropriate time.
+// -----------------------------------------------------------------------------
+
+// FINAL LOGGING SUPPORT FOR PREVIOUS TEST CASE
+// In an ideal world this code would be in an afterEach call, however, since 
+// there will be other afterEach functions that will be called after ours and 
+// we want to ensure the logging we do here is the final logging for the test 
+// case.
+//
+// Also important to notice is that this results in a call to cy.writeFile
+// which will execute in a different time sequence than all of the other 
+// JavaScript in this code block. Thus in this code block we cannot run any
+// true `beforeEach` code that is related to the current test case that is
+// just about to start.
+// FINAL LOGGING SUPPORT FOR PREVIOUS TEST CASE
+beforeEach(function() {
+  LogTestState()
+  FlushLogEntries()
+})
+
+// TRUE `beforeEach` CODE FOR THE TEST CASE THAT IS JUST NOW STARTING
+// This MUST come after the prior `beforeEach` code block and cannot
+// be combined with that other code block due to the `cy.writeFile`
+// call that it contains.
+beforeEach(function() {
+  console.log('')
+  console.log(`******************** ${GetFullTestSuiteTitle(this.currentTest)} ************************************************************`)
+
+  test = this.currentTest
+  let title = GetSuiteRootTitle(this.currentTest)
```

Comment:
This function could return arbitrarily long string or strings with spaces. At the very least it wouldn't be a good candidate for filenames.  

On larger point can you explain a bit more about these logs:
What do these capture that Cypress doesn't capture?
Do these log relate: https://docs.cypress.io/api/events/catalog-of-events.html#Logging-All-Events
Is this just temporary until cypress adds the feature to save logs?

Reply:
Yes these logs are temporary...unless Cypress' future implementation does not work for us. Even in that case we can probaly work through it.

Persistent Logs are for the console.log output. Here are some examples:
> 
-+- 00:50:39..744 - cy.Get([data-testid="train-dialogs-turns"]) - Start - Last DOM change was 787 milliseconds ago
-+- 00:50:39..776 - cy.Get([data-testid="train-dialogs-turns"]) - DOM Is Stable
> 
The above log tells me what UI element we are getting and also shows timing of when the DOM changes, so if we somehow miss something I either need to adjust the algorithm (which has not happened in months) or there is some funky UI rendering going on that I need to file a bug on.

> 
-+- 00:50:39..816 - StringArrayFromElementText([data-testid="train-dialogs-first-input"]) - 1
-+- 00:50:39..816 - StringArrayFromElementText([data-testid="train-dialogs-first-input"]) - Hello
> 
These logs are from a part of the code that often fails. Typically when making a mistake calling the functions using this, but also when unexpected UI elements are present like newlines or that error icon. So this shows what exactly was found so that test code can use the appropriate values.

Reply:
As for the length of the root title...

I believe:
1. We should keep these as short as possible. 
2. We should name our spec file with this same title.
3. Even if it is long, the file system can handle it.

I can and should remove the spaces and camel case it. Given my beliefs about this, I think it makes a perfect name for the logs, makes it easy to figure out which log goes with which test case. I've used this logging mechanism on every project I've been on that I have designed the logging on (all 3 of them) and nobody else complained about it. I've seen others use it in a similar way too.

Reply:
> We should keep these as short as possible.

More important to be meaningful than short.

> Even if it is long, the file system can handle it.

Hmm, I thought there were versions of Windows that still the 260 max path limit.  I thought if the developer cloned the repo in a location relatively deep and we try to write a file which adds additional characters we might reach this limit and fail.  I didn't want to risk someone not being able to run the tests simply be cause they cloned repo in certain location or if they wrote a long test name. In any case this will probably be rare.

> I can and should remove the spaces and camel case it

Were you intending to do this? Or leaving it?

I had seen this line:
https://github.com/Microsoft/ConversationLearner-UI/pull/1020/files#diff-e20a789bb1cfa9903bd0f7f355cc850dR62

which does some concatenation with periods for other parts like `YY.MM.DD.HH.mm.ss..SSS` and thought you were doing it to prevent spaces.

If it can't fail then I don't care as much about the name.

Reply:
Actually it is nice with the spaces in, so I would rather leave it.

As for the periods separating year, month, day, etc...I like having them as separators as it helps make sense of what you are looking at when trying to distinguish one log from another with the same root name.

I spent about an hour trying to make it not fail if someone has a path that is too long, could not do it. But I really don't think we should just throw this functionality out if someone makes their root path too long and without writing the logs to a file this entire code file is useless.

In my experience with test development, most developers understand this restriction and don't do that, and those that do spend a few hours getting a lesson in what not to do that prevents them from repeating that mistake in the future. Such is the life and learnings of a developer.